temporalio/client.py

```diff
@@ -1984,114 +2095,1656 @@ async def __anext__(self) -> temporalio.api.history.v1.HistoryEvent:
+            else None,
+            jitter=spec.jitter.ToTimedelta() if spec.HasField("jitter") else None,
+            time_zone_name=spec.timezone_name or None,
+        )
+
+    def _to_proto(self) -> temporalio.api.schedule.v1.ScheduleSpec:
+        start_time: Optional[google.protobuf.timestamp_pb2.Timestamp] = None
+        if self.start_at:
+            start_time = google.protobuf.timestamp_pb2.Timestamp()
+            start_time.FromDatetime(self.start_at)
+        end_time: Optional[google.protobuf.timestamp_pb2.Timestamp] = None
+        if self.end_at:
+            end_time = google.protobuf.timestamp_pb2.Timestamp()
+            end_time.FromDatetime(self.end_at)
+        jitter: Optional[google.protobuf.duration_pb2.Duration] = None
+        if self.jitter:
+            jitter = google.protobuf.duration_pb2.Duration()
+            jitter.FromTimedelta(self.jitter)
+        return temporalio.api.schedule.v1.ScheduleSpec(
+            structured_calendar=[cal._to_proto() for cal in self.calendars],
+            cron_string=self.cron_expressions,
+            interval=[i._to_proto() for i in self.intervals],
+            exclude_structured_calendar=[cal._to_proto() for cal in self.skip],
+            start_time=start_time,
+            end_time=end_time,
+            jitter=jitter,
+            timezone_name=self.time_zone_name or "",
+        )
+
+
+@dataclass(frozen=True)
+class ScheduleRange:
+    """Inclusive range for a schedule match value.
+
+    .. warning::
+        Schedules are an experimental feature.
+    """
+
+    start: int
+    """Inclusive start of the range."""
+
+    end: int = 0
+    """Inclusive end of the range.
+    
+    If unset or less than start, defaults to start.
+    """
+
+    step: int = 0
+    """
+    Step to take between each value.
+
+    Unset or 0 defaults as 1.
+    """
+
+    @staticmethod
+    def _from_protos(
+        ranges: Sequence[temporalio.api.schedule.v1.Range],
+    ) -> Sequence[ScheduleRange]:
+        return tuple(ScheduleRange._from_proto(r) for r in ranges)
+
+    @staticmethod
+    def _from_proto(range: temporalio.api.schedule.v1.Range) -> ScheduleRange:
+        return ScheduleRange(start=range.start, end=range.end, step=range.step)
+
+    @staticmethod
+    def _to_protos(
+        ranges: Sequence[ScheduleRange],
+    ) -> Sequence[temporalio.api.schedule.v1.Range]:
+        return tuple(r._to_proto() for r in ranges)
+
+    def _to_proto(self) -> temporalio.api.schedule.v1.Range:
+        return temporalio.api.schedule.v1.Range(
+            start=self.start, end=self.end, step=self.step
+        )
+
+
+@dataclass
+class ScheduleCalendarSpec:
+    """Specification relative to calendar time when to run an action.
+
+    A timestamp matches if at least one range of each field matches except for
+    year. If year is missing, that means all years match. For all fields besides
+    year, at least one range must be present to match anything.
```

Comment:
>at least one range must be present to match anything

Could we reword this? I'd argue this is confusing now all ranges have defaults, but it is technically correct I guess

Reply:
Sure, suggestions?

Reply:
thinking about it again I can't think of a better way to word it.